File: 3_pointer_aliasing/cpp/1/main.cpp
#include <iostream>
#include <memory>
#include "main.hpp"

using namespace std;

int main(void) {
    funcA();
    funcB();
    funcC();
}

// Demonstrates the necessity of explicit move semantics
void funcA(void) {
    auto foo = make_unique<string>("bar");
    cout << "funcA(1)" << *foo << endl; 
    /* Uncommenting final line in this function gives this (intended) compilation error:
     * "error: call to implicitly-deleted copy constructor of ...unique_ptr...
     *  note: copy constructor is implicitly deleted because ...unique_ptr... has a
     *  user-declared move constructor."
     */
    //take_ownership(foo);
}

// Demonstrates run-time errors for using moved values
void funcB(void) {
    auto foo = make_unique<string>("bar");
    cout << "funcB(1): " << *foo << endl; 
    take_ownership(move(foo));
    /* Uncommenting the following line will create a runtime error for 
     * dereferencing nullptr.
     */
    //cout << "funcB(2): " << *foo << endl;
}

// Demonstrates nullptr for moved values
void funcC(void) {
    auto foo = make_unique<string>("bar");
    cout << *foo << endl; 
    take_ownership(move(foo));
    if ( nullptr == foo ) {
        cout << "foo is null now" << endl;
    }
}

template <typename T>
inline void take_ownership(T baz) { }

File: 3_pointer_aliasing/rust/1/src/main.rs
fn main() {
    func_a();
    func_b();
    func_c();
}

// Copying
fn func_a() {
    let foo = String::from("bar");
    println!("func_a(1): {}", foo);
    take_ownership(foo.clone());
    println!("func_a(2): {}", foo);
}

fn func_b() {
    let foo = String::from("bar");
    println!("func_b(1): {}", foo);
    let qux = return_ownership(foo);
    println!("func_b(2): {}", qux);
}

fn func_c() {
    let foo = String::from("bar");
    println!("func_c(1): {}", foo);
    borrow_ownership(&foo);
    println!("func_c(2): {}", foo);
}

fn take_ownership<T>(_: T) { }
fn return_ownership<T>(baz: T) -> T { return baz; }
fn borrow_ownership<T>(_: &T) { }